### DISCUSSION TITLE
Datetimes with timezones whose utc-equivalent is BC dates cannot be deserialized back into datetimes

### DISCUSSION THREAD

**Discussion by Ran4** [sqlalchemy/sqlalchemy#6037] [Usage Questions] [answered]
**Describe the bug**
If we create a datetime at year 0, month 1, day 1 with a timezone such that the utc equivalent
would refer to a UTC time that is in a BC year a and save it to a database,
then we can't deserialize it back to a datetime - as when we write the utc

**Expected behavior**

We shouldn't be able to insert datetime values that can't be deserialized back into a datetime.

**To Reproduce**

```py
from uuid import uuid4, UUID
from datetime import datetime, timezone, timedelta

from sqlalchemy import (
    Column,
    DateTime,
)
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


class Example(Base):
    __tablename__ = "example"

    example_id = Column(UUID(as_uuid=True), primary_key=True)
    date = Column(DateTime(timezone=True), nullable=True)


bc_datetime = datetime(year=1, month=1, day=1, tzinfo=timezone(timedelta(hours=1)))

db_session.add(
    Example(
        example_id=uuid4(),
        date=bc_datetime.
    )
)
db.commit()

# This works

# but this won't, as date can't be serialized into a datetime:

example = db.query(Example).first()
example.date
```

**Versions.**
 - OS: Ubuntu 20.04.2 LTS
 - Python: 3.8.5
 - SQLAlchemy:
 - Database: PostgreSQL 11.9
 - DBAPI:

**Comment by CaselIT** [comment]
Hi,

Could you also share the stack-trace that is raised?
Also which version of sqlalchemy are you using?

Thanks

**Comment by CaselIT** [comment]
I cannot reproduce with sqlalchemy 1.3.23 and psycopg2 2.8.6

Please adapt the following script to reproduce the error:
```py
from datetime import datetime, timezone, timedelta

from sqlalchemy import Column, DateTime, Integer
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Example(Base):
    __tablename__ = "example"
    example_id = Column(Integer, primary_key=True)
    date = Column(DateTime(timezone=True), nullable=True)

bc_datetime = datetime(year=1, month=1, day=1, tzinfo=timezone(timedelta(hours=1)))
import sqlalchemy as sa

engine = sa.create_engine("postgresql://scott:tiger@localhost/test")

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)
session = sa.orm.Session(engine)

session.add(Example(date=bc_datetime))
session.commit()
session.close()
example = session.query(Example).first()
print(example.date == bc_datetime)
```

**Comment by Ran4** [comment]
I'm using pcycopg2 2.8.6
sqlalchemy 1.3.20

Running the code you posted:

```py
from datetime import datetime, timezone, timedelta

from sqlalchemy import Column, DateTime, Integer
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Example(Base):
    __tablename__ = "example"
    example_id = Column(Integer, primary_key=True)
    date = Column(DateTime(timezone=True), nullable=True)

bc_datetime = datetime(year=1, month=1, day=1, tzinfo=timezone(timedelta(hours=1)))
import sqlalchemy as sa

DB_URL = "postgresql+psycopg2://postgres:mysecretpassword@localhost:9432/db"
engine = sa.create_engine(DB_URL)

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)
session = sa.orm.Session(engine)

session.add(Example(date=bc_datetime))
session.commit()
session.close()
example = session.query(Example).first()
print(example.date == bc_datetime)
```

I get this traceback:

```py
Traceback (most recent call last):
  File "reproduce.py", line 26, in <module>
    example = session.query(Example).first()
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/orm/query.py", line 3429, in first
    ret = list(self[0:1])
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/orm/query.py", line 3203, in __getitem__
    return list(res)
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/orm/loading.py", line 100, in instances
    cursor.close()
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/orm/loading.py", line 76, in instances
    fetch = cursor.fetchall()
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/engine/result.py", line 1288, in fetchall
    self.connection._handle_dbapi_exception(
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1514, in _handle_dbapi_exception
    util.raise_(exc_info[1], with_traceback=exc_info[2])
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/engine/result.py", line 1284, in fetchall
    l = self.process_rows(self._fetchall_impl())
  File "/home/ran/other/example/.venv/lib/python3.8/site-packages/sqlalchemy/engine/result.py", line 1230, in _fetchall_impl
    return self.cursor.fetchall()
ValueError: year 0 is out of range
```

Note: My system is set to CET (+0100). This might be relevant?
I'm running the postgres:11.9 docker image, where I haven't changed the timezone:

```
db=# show timezone
 TimeZone
----------
 Etc/UTC
(1 row)
```

**Comment by CaselIT** [comment, marked as answer]
I'm also in cet, Using docker I can reproduce. I originally tried using a postgres installed in my local system and it does not presented the issue since I guess the returned time was still in cet.

This is an issue of the python datetime library, since `datetime(0, 1, 1)` will error.

If you need to work with year 0 or lower I think you should probably cast the returned value to string and then convert that in your code as you see fit. You can use a custom type for this, like
```py

class StrDateTime(sa.types.TypeDecorator):
    impl = DateTime

    def column_expression(self, value):
        return sa.cast(value, sa.Text)
```

So the full example becomes
```py
from datetime import datetime, timezone, timedelta

import sqlalchemy as sa
from sqlalchemy import Column, DateTime, Integer
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


class StrDateTime(sa.types.TypeDecorator):
    impl = DateTime

    def column_expression(self, value):
        return sa.cast(value, sa.Text)

class Example(Base):
    __tablename__ = "example"
    example_id = Column(Integer, primary_key=True)
    date = Column(StrDateTime(timezone=True), nullable=True)

engine = sa.create_engine("postgresql://scott:tiger@localhost/test", echo=True)

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)
session = sa.orm.Session(engine)

session.add(Example(date='1001-12-31 BC'))
session.commit()
session.close()
example = session.query(Example).first()
print(example.date)
```

@zzzeek do you have other suggestions?

**Comment by Ran4** [comment]
My application doesn't need BC dates, so I solved it by verifying that the datetime instance is valid when converted to utc (    checking that `dt.astimezone(tz=None)` doesn't raise an error).

But I'm thinking that perhaps SQLAlchemy should refuse to write datetimes that aren't valid to read back when writing to Postgres.

I guess it gets a bit more complicated since if you create a `datetime(year=1, month=1, day=1)` with timezone utc+01` and save it to a postgres configured to utc+0, then you'd be unable to read back the postgres datetime into a python datetime UNLESS the postgres instance changed the datetime *after* you wrote the data. To avoid this issue you'd have to disallow writing any datetimes that are still legal when converted from any time zone.

**Comment by CaselIT** [comment]
The date is valid in postgresql so it would be wrong for sqlalchemy to prevent adding that value. Also the library in general does not validate the input data, but will let the dbapi / db do that, since it would be an enormous amount of mostly duplicated work.

In this case I would place the blame on psycopg2 that maps python datetime object to postgres type directly even if the range of postgresql is grater than python ones. I'm not sure if you can change this behaviour.

Other dbapi may choose to treat it in a different manner.

Regarding you application, if you are trying to create a "no-date" date, I personally think that null would be better for that

**Comment by Ran4** [comment]
Yeah, perhaps this should be on psycopg2? I will admit that I don't know exactly who does what, and perhaps this isn't primarily a SQLAlchemy issue.

I would argue that this isn't just a super niche edge case (surely the vast majority of sqlalchemy users don't need BC support!), but it is a possible security risk: a malicious user can post an "invalid" datetime (as in, a python datetime that when saved to a postgres database cannot be read back to a datetime), and once that gets into the database then any sqlalchemy queries that tries to retrieve these rows (serializing into a python datetime) would fail, possibly hosing the entire service.

...this exact thing happened at my company :)

**Comment by CaselIT** [comment]
>Yeah, perhaps this should be on psycopg2? I will admit that I don't know exactly who does what, and perhaps this isn't primarily a SQLAlchemy issue.

It may be worth checking psycopg2 documentation and/or file an issue.
This is an example that uses only psycopg2:
```py
import psycopg2

conn = psycopg2.connect("postgresql://scott:tiger@localhost/test")
cursor = conn.cursor()
cursor.execute("select '0001-01-01T00:00:00+01:00'::TIMESTAMP WITH TIME ZONE")
print(cursor.fetchone())
```
that errors with `ValueError: year 0 is out of range`

**Comment by CaselIT** [comment]
> ...this exact thing happened at my company :)

Interesing. I was also not aware of this limitation of the python library. Will file a bug at my company :)

**Comment by zzzeek** [comment]
Well finally here's a good reason not to use TIMESTAMP WITH TIMEZONE.   The PostgreSQL community swears by this datatype because it's "better" but I always find it to be more confusing than it's worth.    TIMESTAMP WITH TIMEZONE is going to be returning dates *by default* in the timezone of the server, but you can of course modify this to return the datetimes in whatever timezone you want by applying the "AT TIME ZONE" operator.   I'm too tired to get the date math to work correctly but this is the general idea:

```
class Example(Base):
    __tablename__ = "example"
    example_id = Column(Integer, primary_key=True)

    date = deferred(Column("date", DateTime(timezone=True), nullable=True))

    date_utc = column_property(cast(date.expression.op("AT TIME ZONE")("-1"), DateTime(timezone=True)))

```

People never ask for the above recipe because nobody really needs it, they just get the datetimes in and out and whether or not it has tzinfo, it just represents a point in time and they use conversion for it to be in the tzinfo they want.

the only way to "validate" this would be to convert the datetime to UTC ahead of time on the client, which would completely defeat the purpose of using the TIMESTAMP WITH TIME ZONE datatype.   definitely out of scope for what the default DateTime object would be doing, and I would even wonder why Python lets us create a datetime object with such a tzinfo, since the UTC time it refers to is not representable.

I'm pretty much 100% sure Python devs will dismiss this issue as "well then check it yourself" but let me just see how they verbalize this, I can also create this magic date using plain old strptime

    d2 = datetime.strptime("Mon Jan  1 00:00:00 0001 +01:00", "%c %z")

and then it can't be converted to utc:

    print(d.astimezone(timezone.utc))
    OverflowError: date value out of range

**Comment by CaselIT** [comment]
Not sure I agree. I guess this is a good reason to not mapTIMESTAMP WITH TIMEZONE to python datetime objects, but I agree that it's in general useful to have them automatically map them.

I think this can be converted to a discussion

**Reply by zzzeek** [reply]
I tried to ask about this here : https://bugs.python.org/issue43484 but they dont seem to have the appetite to treat it like a bug in Python.

**Reply by CaselIT** [reply]
should we ask on psycopg2?

**Reply by CaselIT** [reply]
well, it's not yet decided, another poster think it's a bug
